experiments/storm/src/main/scala/io/gearpump/experiments/storm/producer/StormSpoutOutputCollector.scala

```diff
@@ -36,7 +45,43 @@ private[storm] class StormSpoutOutputCollector(collector: StormOutputCollector)
     throw throwable
   }
 
-  override def emitDirect(taskId: Int, streamId: String, values: JList[AnyRef], messageId: scala.Any): Unit = {
+  override def emitDirect(taskId: Int, streamId: String, values: JList[AnyRef], messageId: Object): Unit = {
+    setPendingOrAck(messageId)
     collector.emitDirect(taskId, streamId, values)
   }
+
+
+  def ackPendingMessage(checkpointClock: TimeStamp): Unit = {
+    pendingMessage.foreach { case PendingMessage(id, messageTime, _) =>
+      if (messageTime <= checkpointClock) {
+        spout.ack(id)
+        reset
+      }
+    }
+  }
+
+  def failPendingMessage(timeout: Int): Unit = {
+    pendingMessage.foreach { case PendingMessage(id, _, startTime) =>
+      if (System.currentTimeMillis() - startTime >= timeout) {
+        spout.fail(id)
+      }
+      reset
+    }
+  }
+
+  private def reset: Unit = {
+    pendingMessage = None
+  }
+
+  private def setPendingOrAck(messageId: Object): Unit = {
+    val curTime = System.currentTimeMillis()
+    if (ackEnabled && pendingMessage.isEmpty) {
+      val startTime = curTime
+      val messageTime = curTime
+      pendingMessage = Some(PendingMessage(messageId, messageTime, startTime))
+    } else {
+      spout.ack(messageId)
```

Comment:
yes. https://github.com/apache/storm/blob/0.9.x-branch/storm-core/src/clj/backtype/storm/daemon/executor.clj#L485. but I got the order wrong. Updated now.